### PR TITLE
fix: support indexing more types

### LINKED ISSUE #29441: [pkg/ottl] Allow indexing any slice type
I was just shown the scenario 

`set(resource.attributes["my.environment.2"], Split(resource.attributes["host.name"],"-")[1])`

that produces an error:

```
{"kind": "processor", "name": "transform", "pipeline": "logs", "error": "type, []string, does not support int indexing", "statement": "set(resource.attributes[\"my.environment.2\"], Split(resource.attributes[\"host.name\"],\"-\")[1])"}
```

Although the language does support indexing the return value of the Converter, it only supports indexing `pcommon.Slice` and `[]any`, but `Split` returns a `[]string`. We need to make indexing slices more generic.  This might need to be its own issue. Related code: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/60430e1f4c2bdb8579fc080d5f3a6c8f15730a4b/pkg/ottl/expression.go#L98-L112

_Originally posted by @TylerHelmuth in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/26108#issuecomment-1821560835_

### PR DESCRIPTION
**Pull request by znd4**
#### Description

Bugfix: support indexing more types

> [!IMPORTANT]  
> I'm not sure if this should support indexing into slices of `byte` or `[]byte` (I've gone with `[]byte` for now

#### Link to tracking issue

Fixes #29441

#### Testing

currently running `make gotest` (it's taking longer than I expected). I noticed that https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35581 added to `e2e_test.go`. I can add a test case there, but it'll take me a bit longer to grok how those are working 

Update: I'm currently getting `make gotest` failures on `main` from `internal/scraper/processscraper`

#### Documentation

changelog

### CODE REVIEW

**Comment by znd4** [comment]
Ah, sorry I didn't get how to run just the `pkg` tests, and one of the other groups was causing `make gotest` to fail early (guessing b.c. I'm on darwin)

**Comment by znd4** [comment]
@TylerHelmuth , successfully passed the unit tests now with `make gotest GROUP=pkg`

File: pkg/ottl/expression.go [hunk 1]
```diff
@@ -120,6 +120,26 @@ func (g exprGetter[K]) Get(ctx context.Context, tCtx K) (any, error) {
...
+				result, err = getElementByIndex(r, k.Int)
+				if err != nil {
+					return nil, err
+				}
+			case []int64:
+				result, err = getElementByIndex(r, k.Int)
+				if err != nil {
+					return nil, err
+				}
+			case [][]byte:
```
Comment (TylerHelmuth) [inline pkg/ottl/expression.go:138]: @znd4 lets handle this as `[]byte`

File: pkg/ottl/expression_test.go [hunk 2]
```diff
@@ -266,6 +306,86 @@ func Test_newGetter(t *testing.T) {
...
 		},
+		{
+			name: "function call nested SliceBool",
+			val: value{
+				Literal: &mathExprLiteral{
+					Converter: &converter{
+						Function: "SliceBool",
+						Keys: []key{
+							{
+								// note for review: not sure if this is correct
```
Comment (evan-bradley) [inline pkg/ottl/expression_test.go:317]: I think this is fine as-is. We're indexing the outer `[]any` slice, then the inner `[]bool` slice, so we have two integer keys here.

Comment (znd4) [reply pkg/ottl/expression_test.go:317]: ohh okay I was struggling to understand what this was for. Not sure if there's a better option than this though -- the joys of abstraction

**Review by TylerHelmuth — APPROVED** [review]

File: .chloggen/index_more_types.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: pkg/ottl
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Allow indexing []int64, []float64, []bool, []byte slices
```
Comment (evan-bradley) [inline .chloggen/index_more_types.yaml:10]:
```suggestion
note: Allow indexing []int64, []float64, []bool, and []byte slices
```

File: .chloggen/index_more_types.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [29441]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
```
Comment (evan-bradley) [inline .chloggen/index_more_types.yaml:18]:
```suggestion
subtext: It should now be possible to index all slice types
```

**Review by evan-bradley — APPROVED** [review]

**Pull request merged**